{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introducing the next generation of InstaNovo models\n",
    "\n",
    "## Announcing InstaNovo v1.1\n",
    "\n",
    "Since our [InstaNovo paper](https://www.nature.com/articles/s42256-025-01019-5) is now published, we’d like to share an update on what we’ve been working on while our manuscript was under review. With the release of our preprint over a year ago, we were overwhelmed by the incredible response from the community and the applications of our models. We are actively collaborating with leading experts and continue to explore the potential of powerful solutions for de novo peptide sequencing. Our ongoing efforts focus on developing more accurate models, expanding the de novo sequencing ecosystem for analysis and data reporting, fine-tuning our models, and designing tailored, application-specific workflows.\n",
    "\n",
    "While much of this work is still in progress, we have just released an improved version of our base model, InstaNovo v1.1. This model boasts higher recall, greater identification certainty, expanded support for modifications, and enhanced data processing and reporting features. We believe these advancements are worth communicating with this post instead of an article, and we are excited to show you how this model compares to the earlier model in our paper.\n",
    "\n",
    "## Getting started\n",
    "\n",
    "Our new model is available in the [main InstaNovo branch](https://github.com/instadeepai/InstaNovo/tree/main) with detailed documentation on installation, local execution, and running it in training or testing mode with your data. If you prefer a hosted solution, InstaDeep provides access to the model via our website, where you can upload your files and analyse up to 100,000 spectra per day on our servers (free for academic use). To reproduce the analysis in this blog post or to run this analysis on your own dataset, you can use this Jupyter notebook.\n",
    "\n",
    "## Training data \n",
    "\n",
    "InstaNovo v1.1 (IN v1.1) has been trained on the [ProteomeTools](https://www.proteometools.org/), [MassiveKB](https://massive.ucsd.edu/ProteoSAFe/static/massive-kb-libraries.jsp), and kind dataset contributions from several other projects processed by the [CompOmics](https://www.compomics.com/) group at Ghent University. We are grateful for the huge community effort to generate, process, and curate the datasets that have enabled the development of our models. We are excited to contribute with the release of this combined dataset in the near future.\n",
    "\n",
    "## Benchmarking IN v1.1\n",
    "\n",
    "[IN v1.1](https://github.com/instadeepai/InstaNovo/releases/tag/1.1.0) is a substantial improvement over [IN v0.1](https://github.com/instadeepai/InstaNovo/releases/tag/0.1.4), the original model from our paper. To benchmark these models, we used a standard HeLa proteome run, a widely used reference sample in proteomics facilities for quality control and one of the validation datasets in our study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "sns.set_theme('paper', 'ticks', 'colorblind', font_scale=1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "# Suppress only FutureWarnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_dict = {'old_knapsack': '#E2DBBE', 'new_knapsack': '#188FA7', 'new_greedy': '#9DBBAE', 'database': '#769FB6'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir ./data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python ../instanovo/scripts/get_zenodo_record.py --zenodo-url https://zenodo.org/api/records/15174703/files-archive --zip-path 15174703.zip --extract-path ./data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final = pd.read_csv('./data/hela_preds.csv')\n",
    "targets = pd.read_csv('./data/hela_targets.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge final and target on scan_number\n",
    "final = final.merge(targets, on='scan_number', how='left', suffixes=('_final', '_target'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalise(string):\n",
    "    string = string.replace(\"UNIMOD\", \"\").replace(\"I\", \"L\")\n",
    "    return ''.join(re.findall(r'[A-Z]', string))\n",
    "\n",
    "\n",
    "# get proteins from dataframe\n",
    "def get_proteins(df, column):\n",
    "    proteins = set()\n",
    "    for ind in df.index:\n",
    "        try:\n",
    "            proteins.add(df.loc[ind, column].split(';')[0].split('|')[1])\n",
    "        except:\n",
    "            print(df.loc[ind, column], ind)\n",
    "            pass\n",
    "    return proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final['preds'] = final['preds'].replace(np.nan, '')\n",
    "final['preds_new_knapsack'] = final['preds_new_knapsack'].replace(np.nan, '')\n",
    "final['preds_new_greedy'] = final['preds_new_greedy'].replace(np.nan, '')\n",
    "final['Sequence'] = final['Sequence'].replace(np.nan, '')\n",
    "final['sequence'] = final['sequence'].replace(np.nan, '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final['stripped_preds_old_knapsack'] = final['preds'].apply(normalise)\n",
    "final['stripped_preds_new_knapsack'] = final['preds_new_knapsack'].apply(normalise)\n",
    "final['stripped_preds_new_greedy'] = final['preds_new_greedy'].apply(normalise)\n",
    "final['stripped_sequence'] = final['Sequence'].apply(normalise)\n",
    "final['stripped_target'] = final['sequence'].apply(normalise)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use peptide-level metrics and especially recall to assess our models, the most direct evaluation of prediction performance. This is because bottom up proteomics is a peptide centric methodology; if the peptide is wrong, it does not matter what percentage of amino acids is wrong. Crucially, we evaluate whether the full prediction is correct.\n",
    "\n",
    "The distribution of model confidence, defined as the product of our residue log probabilities raised in its natural exponent, indicates that the new model is sharper in its prediction certainty. We observe more high confidence predictions being correct, while lower confidence predictions are more densely clustered in the lower confidence range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's start by plotting the confidence range as a kernel density plot for old and new knapsack\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.kdeplot(final['confidence_old_knapsack'], color=color_dict['old_knapsack'], label='Old Knapsack')\n",
    "sns.kdeplot(final['confidence_new_knapsack'], color=color_dict['new_knapsack'], label='New Knapsack')\n",
    "plt.xlabel('Confidence')\n",
    "plt.ylabel('Density')\n",
    "plt.legend(['IN v0.1 knapsack', 'IN v1.1 knapsack'], loc='upper right')\n",
    "plt.title('Model confidence range')\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's do a stacked bar plot instead\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.histplot(final[['confidence_old_knapsack', 'confidence_new_knapsack']], bins=20, kde=False, multiple='stack', palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.xlabel('Confidence')\n",
    "plt.ylabel('Density')\n",
    "plt.legend(['IN v0 knapsack', 'IN v1.1 knapsack', 'IN 1.1 greedy'])\n",
    "plt.title('Model confidence range')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's do overlaid histograms instead\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.histplot(final['confidence_old_knapsack'], bins=20, kde=False, color=color_dict['old_knapsack'], alpha=0.5, label='Old Knapsack')\n",
    "sns.histplot(final['confidence_new_knapsack'], bins=20, kde=False, color=color_dict['new_knapsack'], alpha=0.5, label='New Knapsack')\n",
    "plt.xlabel('Confidence')\n",
    "plt.ylabel('Density')\n",
    "plt.legend(['IN v0 knapsack', 'IN v1.1 knapsack', 'IN 1.1 greedy'])\n",
    "plt.title('Model confidence range')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's do all three now with kde\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.kdeplot(final['confidence_old_knapsack'], color=color_dict['old_knapsack'], label='Old Knapsack', alpha=0.5)\n",
    "sns.kdeplot(final['confidence_new_knapsack'], color=color_dict['new_knapsack'], label='New Knapsack', alpha=0.5)\n",
    "sns.kdeplot(final['confidence_new_greedy'], color=color_dict['new_greedy'], label='New Greedy', alpha=0.5, linestyle='--', linewidth=3)\n",
    "plt.xlabel('Confidence')\n",
    "plt.ylabel('Density')\n",
    "plt.legend()\n",
    "plt.title('Model confidence range, density plot')\n",
    "plt.legend(['IN v0.1 knapsack', 'IN v1.1 knapsack', 'IN 1.1 greedy'])\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all three with stacked bar plot\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.histplot(final[['confidence_old_knapsack', 'confidence_new_knapsack', 'confidence_new_greedy']], bins=20, kde=False, multiple='stack', palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.xlabel('Confidence')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Model confidence range')\n",
    "plt.legend(['IN v0 knapsack', 'IN v1.1 knapsack', 'IN 1.1 greedy'])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict analysis to rows with annotated database results. Let's use the Confidence column for this, need to remove NA values\n",
    "final_db = final[final['stripped_target'] != '']\n",
    "final_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute recall for old knapsack. Check columns 'preds' and 'Sequence' for exact match. Recall would be number of exact matches divided by total number of rows\n",
    "recall_old_knapsack = (final_db['stripped_preds_old_knapsack'] == final_db['stripped_target']).sum() / final_db.shape[0]\n",
    "print(f'Recall for old knapsack: {recall_old_knapsack:.4f}')\n",
    "\n",
    "# do the same for new knapsack\n",
    "recall_new_knapsack = (final_db['stripped_preds_new_knapsack'] == final_db['stripped_target']).sum() / final_db.shape[0]\n",
    "print(f'Recall for new knapsack: {recall_new_knapsack:.4f}')\n",
    "\n",
    "# and for new greedy\n",
    "recall_new_greedy = (final_db['stripped_preds_new_greedy'] == final_db['stripped_target']).sum() / final_db.shape[0]\n",
    "print(f'Recall for new greedy: {recall_new_greedy:.4f}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compared to the gold standard database search, IN v0.1 achieved a recall of 49.5%. IN v1.1, using greedy precursor mass fit search (i.e. selecting the sequence that best fits the observed precursor mass), improves recall to 56.7%, while knapsack beam search further boosts recall to 63%. This compounds to a 13.5% improvement over the previous model, with more than six out of ten peptides being detected without precursor mass filtering on predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot bar plot for recall, old knapsack, new knapsack, new greedy\n",
    "plt.figure(figsize=(6, 5))\n",
    "sns.barplot(x=['IN v0.1 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[recall_old_knapsack, recall_new_knapsack, recall_new_greedy], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.ylabel('Recall')\n",
    "# put numbers on top of bars\n",
    "for i, v in enumerate([recall_old_knapsack, recall_new_knapsack, recall_new_greedy]):\n",
    "    plt.text(i, v + 0.0005, f'{v:.2f}', ha='center')\n",
    "plt.title('Model recall compared to database')\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot bar plot for recall, old knapsack and new knapsack\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v0', 'IN v1.1'], y=[recall_old_knapsack, recall_new_knapsack], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.ylabel('Recall')\n",
    "# put numbers on top of bars\n",
    "for i, v in enumerate([recall_old_knapsack, recall_new_knapsack]):\n",
    "    plt.text(i, v + 0.0005, f'{v:.2f}', ha='center', va='bottom')\n",
    "\n",
    "# adjust space between plot and title\n",
    "plt.subplots_adjust(top=0.9)\n",
    "plt.title('Recall for old vs new knapsack')\n",
    "plt.ylim(0, 0.7)\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=0.9)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute precision vs recall curve for old knapsack\n",
    "precision_old_knapsack = []\n",
    "recall_old_knapsack = []\n",
    "confidence_old_knapsack = []\n",
    "\n",
    "for threshold in np.linspace(0, 1, 100):\n",
    "    tp = ((final_db['confidence_old_knapsack'] > threshold) & (final_db['stripped_preds_old_knapsack'] == final_db['stripped_target'])).sum()\n",
    "    fp = ((final_db['confidence_old_knapsack'] > threshold) & (final_db['stripped_preds_old_knapsack'] != final_db['stripped_target'])).sum()\n",
    "    fn = ((final_db['confidence_old_knapsack'] <= threshold) & (final_db['stripped_preds_old_knapsack'] == final_db['stripped_target'])).sum()\n",
    "    precision_old_knapsack.append(tp / (tp + fp))\n",
    "    recall_old_knapsack.append(tp / (tp + fn))\n",
    "    confidence_old_knapsack.append(threshold)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute precision vs recall curve for new knapsack\n",
    "precision_new_knapsack = []\n",
    "recall_new_knapsack = []\n",
    "confidence_new_knapsack = []\n",
    "\n",
    "for threshold in np.linspace(0, 1, 100):\n",
    "    tp = ((final_db['confidence_new_knapsack'] > threshold) & (final_db['stripped_preds_new_knapsack'] == final_db['stripped_target'])).sum()\n",
    "    fp = ((final_db['confidence_new_knapsack'] > threshold) & (final_db['stripped_preds_new_knapsack'] != final_db['stripped_target'])).sum()\n",
    "    fn = ((final_db['confidence_new_knapsack'] <= threshold) & (final_db['stripped_preds_new_knapsack'] == final_db['stripped_target'])).sum()\n",
    "    precision_new_knapsack.append(tp / (tp + fp))\n",
    "    recall_new_knapsack.append(tp / (tp + fn))\n",
    "    confidence_new_knapsack.append(threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute precision vs recall curve for new greedy\n",
    "precision_new_greedy = []\n",
    "recall_new_greedy = []\n",
    "confidence_new_greedy = []\n",
    "\n",
    "for threshold in np.linspace(0, 1, 100):\n",
    "    tp = ((final_db['confidence_new_greedy'] > threshold) & (final_db['stripped_preds_new_greedy'] == final_db['stripped_target'])).sum()\n",
    "    fp = ((final_db['confidence_new_greedy'] > threshold) & (final_db['stripped_preds_new_greedy'] != final_db['stripped_target'])).sum()\n",
    "    fn = ((final_db['confidence_new_greedy'] <= threshold) & (final_db['stripped_preds_new_greedy'] == final_db['stripped_target'])).sum()\n",
    "    precision_new_greedy.append(tp / (tp + fp))\n",
    "    recall_new_greedy.append(tp / (tp + fn))\n",
    "    confidence_new_greedy.append(threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We observe better precision values in parallel with increased coverage "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot precision vs recall curve for old knapsack, new knapsack, new greedy\n",
    "plt.figure(figsize=(5, 5))\n",
    "plt.plot(recall_old_knapsack, precision_old_knapsack, label='IN v0.1 knapsack', color=color_dict['old_knapsack'], linewidth=3, linestyle='--')  \n",
    "plt.plot(recall_new_knapsack, precision_new_knapsack, label='IN v1.1 knapsack', color=color_dict['new_knapsack'], linewidth=3)\n",
    "plt.plot(recall_new_greedy, precision_new_greedy, label='IN v1.1 greedy', color=color_dict['new_greedy'], linewidth=3)\n",
    "\n",
    "plt.xlabel('Coverage')\n",
    "plt.ylabel('Precision')\n",
    "plt.legend()\n",
    "plt.title('Precision vs coverage curve')\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot precision vs recall curve for old knapsack, new knapsack\n",
    "plt.figure(figsize=(5, 5))\n",
    "plt.plot(recall_old_knapsack, precision_old_knapsack, label='IN v0 knapsack', color=color_dict['old_knapsack'], linewidth=3, linestyle='--')\n",
    "plt.plot(recall_new_knapsack, precision_new_knapsack, label='IN v1.1 knapsack', color=color_dict['new_knapsack'], linewidth=3)\n",
    "\n",
    "plt.xlabel('Coverage')\n",
    "plt.ylabel('Precision')\n",
    "plt.legend()\n",
    "plt.title('Precision vs coverage curve for old and new knapsack')\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate confidence threshold for 95% precision for old knapsack, new knapsack, new greedy\n",
    "confidence_threshold_old_knapsack = confidence_old_knapsack[np.argmax(np.array(precision_old_knapsack) >= 0.95)]\n",
    "print(f'Old Knapsack: {confidence_threshold_old_knapsack:.2f}')\n",
    "\n",
    "confidence_threshold_new_knapsack = confidence_new_knapsack[np.argmax(np.array(precision_new_knapsack) >= 0.95)]\n",
    "print(f'New Knapsack: {confidence_threshold_new_knapsack:.2f}')\n",
    "\n",
    "confidence_threshold_new_greedy = confidence_new_greedy[np.argmax(np.array(precision_new_greedy) >= 0.95)]\n",
    "print(f'New Greedy: {confidence_threshold_new_greedy:.2f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot confidence threshold for 95% precision for old knapsack, new knapsack, new greedy\n",
    "plt.figure(figsize=(6, 5))\n",
    "plt.bar(['IN v0 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], [confidence_threshold_old_knapsack, confidence_threshold_new_knapsack, confidence_threshold_new_greedy], color=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.ylabel('Confidence threshold for 95% precision')\n",
    "plt.title('Confidence threshold for 95% precision for old and new knapsack, and new greedy')\n",
    "# plot numbers on top of bars\n",
    "for i, v in enumerate([confidence_threshold_old_knapsack, confidence_threshold_new_knapsack, confidence_threshold_new_greedy]):\n",
    "    plt.text(i, v + 0.01, f'{v:.2f}', ha='center')\n",
    "    \n",
    "plt.ylim(0, 1)\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While it is still unclear whether this trend will persist with larger training datasets and increasing model sizes, this pattern suggests that confidence-based thresholding can help reduce false positive rates.\n",
    "\n",
    "We use this precision to estimate the false discovery rate (FDR) and derive confidence cutoffs for low false identifications, which yield identification results aligned with conventional proteomics search outputs. In agreement with our confidence analysis, we observe that lower confidence thresholds are required to maintain a 5% FDR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same plot but only for old and new knapsack\n",
    "plt.figure(figsize=(6, 8))\n",
    "plt.bar(['IN v0.1', 'IN v1.1'], [confidence_threshold_old_knapsack, confidence_threshold_new_knapsack], color=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.ylabel('Confidence range')\n",
    "plt.title('Confidence threshold for 95% precision, knapsack')\n",
    "# plot numbers on top of bars\n",
    "for i, v in enumerate([confidence_threshold_old_knapsack, confidence_threshold_new_knapsack]):\n",
    "    plt.text(i, v + 0.005, f'{v:.2f}', ha='center', va='bottom')\n",
    "plt.ylim(0, 1)\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot barplot of new PSMs at 5% FDR\n",
    "old_ks = final[final['confidence_old_knapsack'] >= confidence_threshold_old_knapsack].shape[0]\n",
    "print(f'Old Knapsack: {old_ks} PSMs at 5% FDR')\n",
    "\n",
    "new_ks = final[final['confidence_new_knapsack'] >= confidence_threshold_new_knapsack].shape[0]\n",
    "print(f'New Knapsack: {new_ks} PSMs at 5% FDR')\n",
    "\n",
    "new_greedy = final[final['confidence_new_greedy'] >= confidence_threshold_new_greedy].shape[0]\n",
    "print(f'New Greedy: {new_greedy} PSMs at 5% FDR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find new PSMs at 5% FDR, i.e. not in the database. Use the Confidence column for this\n",
    "new_ks_fdr = final[(final['confidence_new_knapsack'] >= confidence_threshold_new_knapsack) & (final['stripped_target'] == '')].shape[0]\n",
    "print(f'New Knapsack: {new_ks_fdr} new PSMs at 5% FDR')\n",
    "\n",
    "new_greedy_fdr = final[(final['confidence_new_greedy'] >= confidence_threshold_new_greedy) & (final['stripped_target'] == '')].shape[0]\n",
    "print(f'New Greedy: {new_greedy_fdr} new PSMs at 5% FDR')\n",
    "\n",
    "old_ks_fdr = final[(final['confidence_old_knapsack'] >= confidence_threshold_old_knapsack) & (final['stripped_target'] == '')].shape[0]\n",
    "print(f'Old Knapsack: {old_ks_fdr} new PSMs at 5% FDR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "categories = ['IN v0 knapsack', 'IN 1.1 knapsack', 'IN 1.1 greedy']\n",
    "values = [old_ks, new_ks, new_greedy]  # Total PSMs\n",
    "shading_values = [old_ks - old_ks_fdr, new_ks - new_ks_fdr, new_greedy - new_greedy_fdr]  # Shaded portion (Database-included PSMs)\n",
    "unshaded_values = [old_ks_fdr, new_ks_fdr, new_greedy_fdr]  # Unshaded portion (FDR-controlled PSMs)\n",
    "colors = [color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']]\n",
    "\n",
    "# Create bar plot\n",
    "plt.figure(figsize=(6, 8))\n",
    "ax = sns.barplot(x=categories, y=values, palette=colors, edgecolor=\"none\")  # Remove default bar edges\n",
    "\n",
    "# Overlay shading at the bottom of each bar\n",
    "bar_width = 0.8  # Width of bars\n",
    "alpha = 0.5  # Transparency level for shading\n",
    "for i, (v, sv, uv) in enumerate(zip(values, shading_values, unshaded_values, strict=True)):\n",
    "    # Shaded portion (Database-included PSMs)\n",
    "    plt.gca().add_patch(plt.Rectangle(\n",
    "        (i - bar_width / 2, 0),  # Bottom-left corner\n",
    "        bar_width,  # Width\n",
    "        sv,  # Height\n",
    "        color='gray', alpha=alpha, label=\"Database-included PSMs\" if i == 0 else \"\"  # Only add label once\n",
    "    ))\n",
    "\n",
    "    # Add text labels inside the bars\n",
    "    plt.text(i, sv / 2, f'{sv}', ha='center', va='center', fontsize=15, color='black')  # Center of shaded portion\n",
    "    plt.text(i, sv + (uv / 2), f'{uv}', ha='center', va='center', fontsize=15, color='black')  # Center of unshaded portion\n",
    "    plt.text(i, v + 100, f'{v}', ha='center', va='bottom', fontsize=15, color='black')  # Top of bar\n",
    "\n",
    "# Formatting\n",
    "plt.ylabel('PSMs at 5% FDR')\n",
    "plt.title('PSMs at 5% FDR')\n",
    "plt.ylim(0, max(values) * 1.15)  # Adjust y-limit to fit labels\n",
    "sns.despine()\n",
    "\n",
    "# Add legend for shading, positioned to the right with a box\n",
    "plt.legend(loc='upper center', frameon=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the same plot without greedy\n",
    "categories = ['IN v0 knapsack', 'IN v1.1 knapsack']\n",
    "values = [old_ks, new_ks]  # Total PSMs\n",
    "shading_values = [old_ks - old_ks_fdr, new_ks - new_ks_fdr]  # Shaded portion (Database-included PSMs)\n",
    "unshaded_values = [old_ks_fdr, new_ks_fdr]  # Unshaded portion (FDR-controlled PSMs)\n",
    "colors = [color_dict['old_knapsack'], color_dict['new_knapsack']]\n",
    "\n",
    "# Create bar plot\n",
    "plt.figure(figsize=(5, 8))\n",
    "ax = sns.barplot(x=categories, y=values, palette=colors, edgecolor=\"none\")  # Remove default bar edges\n",
    "\n",
    "# Overlay shading at the bottom of each bar\n",
    "bar_width = 0.8  # Width of bars\n",
    "alpha = 0.5  # Transparency level for shading\n",
    "\n",
    "for i, (v, sv, uv) in enumerate(zip(values, shading_values, unshaded_values, strict=True)):\n",
    "    # Shaded portion (Database-included PSMs)\n",
    "    plt.gca().add_patch(plt.Rectangle(\n",
    "        (i - bar_width / 2, 0),  # Bottom-left corner\n",
    "        bar_width,  # Width\n",
    "        sv,  # Height\n",
    "        color='#171a18', alpha=alpha, label=\"Database-included PSMs\" if i == 0 else \"\"  # Only add label once\n",
    "    ))\n",
    "    \n",
    "    # Add text labels inside the bars\n",
    "    plt.text(i, sv / 2, f'{sv}', ha='center', va='center', fontsize=15, color='white')  # Center of shaded portion\n",
    "    plt.text(i, sv + (uv / 2), f'{uv}', ha='center', va='center', fontsize=15, color='white')  # Center of unshaded portion\n",
    "    plt.text(i, v + 100, f'{v}', ha='center', va='bottom', fontsize=15, color='white')  # Top of bar\n",
    "\n",
    "# Formatting\n",
    "plt.ylabel('PSMs at 5% FDR')\n",
    "plt.title('PSMs at 5% FDR')\n",
    "plt.ylim(0, max(values) * 1.15)  # Adjust y-limit to fit labels\n",
    "sns.despine()\n",
    "\n",
    "# Add legend for shading, positioned to the right with a box\n",
    "plt.legend(loc='upper center', frameon=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the same threshold, IN v1.1 identifies more novel PSMs compared to its predecessor "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the same plot without greedy\n",
    "categories = ['IN v0 knapsack', 'IN v1.1 knapsack']\n",
    "values = [old_ks, new_ks]  # Total PSMs\n",
    "target_values = [old_ks - old_ks_fdr, new_ks - new_ks_fdr]  # Shaded portion (Database-included PSMs)\n",
    "new_values = [old_ks_fdr, new_ks_fdr]  # Unshaded portion (FDR-controlled PSMs)\n",
    "colors = ['#c4b780', '#0f5563', color_dict['old_knapsack'], color_dict['new_knapsack']]\n",
    "\n",
    "# Create stacked bar plot with new values on top of target values for each category\n",
    "plt.figure(figsize=(7, 8))\n",
    "plt.bar(categories, target_values, color=colors[0:2], edgecolor=\"none\", label=\"Database-included PSMs\")\n",
    "\n",
    "plt.bar(categories, new_values, color=colors[2:], edgecolor=\"none\", label=\"New PSMs\", bottom=target_values)\n",
    "\n",
    "# Add text labels inside the bars\n",
    "for i, (nv, tv) in enumerate(zip(new_values, target_values, strict=True)):\n",
    "    plt.text(i, tv / 2, f'{tv}', ha='center', va='center', fontsize=15, color='white')  # Center of target portion\n",
    "    plt.text(i, tv + (nv / 2), f'{nv}', ha='center', va='center', fontsize=15, color='white')  # Center of new portion\n",
    "\n",
    "# Formatting\n",
    "plt.ylabel('PSMs at 5% FDR')\n",
    "plt.title('PSMs at 5% FDR, knapsack')\n",
    "\n",
    "# add all 4 labels and colors to legend\n",
    "# Add custom legend\n",
    "legend_patches = [\n",
    "    mpatches.Patch(color='#c4b780', label='Database shared, IN v0.1'),\n",
    "    mpatches.Patch(color=color_dict['old_knapsack'], label='Novel PSMs, IN v0.1'),\n",
    "    mpatches.Patch(color='#0f5563', label='Database shared, IN v1.1'),\n",
    "    mpatches.Patch(color=color_dict['new_knapsack'], label='Novel PSMs, v1.1')\n",
    "]\n",
    "\n",
    "plt.legend(handles=legend_patches, loc='upper left', frameon=True)\n",
    "\n",
    "plt.ylim(0, max(values) * 1.15)  # Adjust y-limit to fit labels\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to proteome\n",
    "\n",
    "IN v1.1 predicts 42.6% more Peptide Spectrum Matches (PSMs) that map to the proteome with exact sequence matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of mapped PSMs for old and new predictions\n",
    "plt.figure(figsize=(6, 6))\n",
    "psms_old_knapsack = len(final[final.protein_id.notna()])\n",
    "psms_new_knapsack = len(final[final.protein_id_new_knapsack.notna()])\n",
    "psms_new_greedy = len(final[final.protein_id_new_greedy.notna()])\n",
    "\n",
    "sns.barplot(x=['IN v0.1 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[psms_old_knapsack, psms_new_knapsack, psms_new_greedy], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.ylabel('Number of mapped PSMs')\n",
    "plt.ylim(max(0, min(0, max(psms_old_knapsack, psms_new_knapsack, psms_new_greedy) * 1.2)))\n",
    "# put numbers on top of bars\n",
    "for i, v in enumerate([psms_old_knapsack, psms_new_knapsack, psms_new_greedy]):\n",
    "    plt.text(i, v + 100, f'{v}', ha='center', va='bottom')\n",
    "plt.title('Number of mapped PSMs from predictions')\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same plot without greedy\n",
    "plt.figure(figsize=(4, 6))\n",
    "psms_old_knapsack = len(final[final.protein_id.notna()])\n",
    "psms_new_knapsack = len(final[final.protein_id_new_knapsack.notna()])\n",
    "\n",
    "sns.barplot(x=['IN v0', 'IN v1.1'], y=[psms_old_knapsack, psms_new_knapsack], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.ylabel('Number of mapped PSMs')\n",
    "plt.ylim(max(0, min(0, max(psms_old_knapsack, psms_new_knapsack) * 1.2)))\n",
    "# put numbers on top of bars\n",
    "for i, v in enumerate([psms_old_knapsack, psms_new_knapsack]):\n",
    "    plt.text(i, v + 100, f'{v}', ha='center', va='bottom')\n",
    "plt.title('Number of mapped PSMs for knapsack versions')\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get mapped numbers for new and old predictions and plot the for different confidence thresholds\n",
    "confidence_thresholds = np.linspace(0, 1, 50)\n",
    "mapped_old_knap = []\n",
    "mapped_new_knap = []\n",
    "mapped_new_greedy = []\n",
    "\n",
    "for threshold in confidence_thresholds:\n",
    "    matched_old_knapsack = final[(final[\"confidence_old_knapsack\"] > threshold) & (final.protein_id.notna())]\n",
    "    matched_new_knapsack = final[(final[\"confidence_new_knapsack\"] > threshold) & (final.protein_id_new_knapsack.notna())]\n",
    "    matched_new_greedy = final[(final[\"confidence_new_greedy\"] > threshold) & (final.protein_id_new_greedy.notna())]\n",
    "\n",
    "    mapped_old_knap.append(matched_old_knapsack.shape[0])\n",
    "    mapped_new_knap.append(matched_new_knapsack.shape[0])\n",
    "    mapped_new_greedy.append(matched_new_greedy.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(confidence_thresholds, mapped_old_knap, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, mapped_new_knap, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.plot(confidence_thresholds, mapped_new_greedy, label='IN v1.1 greedy', color=color_dict['new_greedy'])\n",
    "plt.scatter(confidence_thresholds, mapped_old_knap, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, mapped_new_knap, color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, mapped_new_greedy, color=color_dict['new_greedy'])\n",
    "plt.ylim(0, max(max(mapped_old_knap), max(mapped_new_knap), max(mapped_new_greedy)) * 1.1)\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Number of mapped predictions')\n",
    "plt.legend()\n",
    "plt.title('Mapped PSMs for different confidence thresholds')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To assess FDR more directly, we mapped predicted PSMs to the human proteome. As expected, we observed an increased proportion of PSMs mapping to the proteome, with the greatest improvements in the high confidence range "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same plot without greedy\n",
    "plt.plot(confidence_thresholds, mapped_old_knap, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, mapped_new_knap, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, mapped_old_knap, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, mapped_new_knap, color=color_dict['new_knapsack'])\n",
    "plt.ylim(0, max(max(mapped_old_knap), max(mapped_new_knap)) * 1.1)\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Number of mapped predictions')\n",
    "plt.legend()\n",
    "plt.title('Mapped PSMs for different confidence thresholds, knapsack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate above code to our data\n",
    "unmapped_old_knap = []\n",
    "unmapped_new_knap = []\n",
    "unmapped_new_greedy = []\n",
    "for threshold in confidence_thresholds:\n",
    "    unmatched_old_knapsack = final[(final[\"confidence_old_knapsack\"] > threshold) & (final.protein_id.isna())]\n",
    "    unmatched_new_knapsack = final[(final[\"confidence_new_knapsack\"] > threshold) & (final.protein_id_new_knapsack.isna())]\n",
    "    unmatched_new_greedy = final[(final[\"confidence_new_greedy\"] > threshold) & (final.protein_id_new_greedy.isna())]\n",
    "    unmapped_old_knap.append(unmatched_old_knapsack.shape[0])\n",
    "    unmapped_new_knap.append(unmatched_new_knapsack.shape[0])\n",
    "    unmapped_new_greedy.append(unmatched_new_greedy.shape[0])\n",
    "\n",
    "plt.plot(confidence_thresholds, unmapped_old_knap, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, unmapped_new_knap, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.plot(confidence_thresholds, unmapped_new_greedy, label='IN v1.1 greedy', color=color_dict['new_greedy'])\n",
    "plt.scatter(confidence_thresholds, unmapped_old_knap, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, unmapped_new_knap, color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, unmapped_new_greedy, color=color_dict['new_greedy'])\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Number of unmapped predictions')\n",
    "plt.legend()\n",
    "plt.title('Unmapped PSMs for different confidence thresholds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same plot without greedy\n",
    "plt.plot(confidence_thresholds, unmapped_old_knap, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, unmapped_new_knap, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, unmapped_old_knap, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, unmapped_new_knap, color=color_dict['new_knapsack'])\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Number of unmapped predictions')\n",
    "plt.legend()\n",
    "\n",
    "plt.title('Unmapped PSMs for different confidence thresholds, knapsack')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get ratio of mapped to unmapped for old knapsack\n",
    "mapped_unmapped_old_knap = np.array(mapped_old_knap) / np.array(unmapped_old_knap)\n",
    "mapped_unmapped_new_knap = np.array(mapped_new_knap) / np.array(unmapped_new_knap)\n",
    "mapped_unmapped_new_greedy = np.array(mapped_new_greedy) / np.array(unmapped_new_greedy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate above code to our data\n",
    "ratio_old = np.array(mapped_old_knap) / np.array(unmapped_old_knap)\n",
    "ratio_new = np.array(mapped_new_knap) / np.array(unmapped_new_knap)\n",
    "ratio_greedy = np.array(mapped_new_greedy) / np.array(unmapped_new_greedy)\n",
    "\n",
    "plt.plot(confidence_thresholds, ratio_old, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, ratio_new, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.plot(confidence_thresholds, ratio_greedy, label='IN v1.1 knapsack greedy', color=color_dict['new_greedy'])\n",
    "plt.scatter(confidence_thresholds, ratio_old, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, ratio_new, color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, ratio_greedy, color=color_dict['new_greedy'])\n",
    "\n",
    "# create a dash line at y ratio of 20\n",
    "plt.axhline(y=20, color='black', linestyle='--')\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Mapped to unmapped ratio')\n",
    "plt.legend()\n",
    "plt.title('Mapped to unmapped ratio for different confidence thresholds')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, when comparing the ratio of mapped to unmapped predictions, IN v1.1 demonstrates a considerable reduction in false positive rates (dashed line indicates 5% FDR when directly mapping to proteome without database grounding)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate above code to our data\n",
    "ratio_old = np.array(mapped_old_knap) / np.array(unmapped_old_knap)\n",
    "ratio_new = np.array(mapped_new_knap) / np.array(unmapped_new_knap)\n",
    "\n",
    "plt.plot(confidence_thresholds, ratio_old, label='IN v0.1 knapsack', color=color_dict['old_knapsack'])\n",
    "plt.plot(confidence_thresholds, ratio_new, label='IN v1.1 knapsack', color=color_dict['new_knapsack'])\n",
    "plt.scatter(confidence_thresholds, ratio_old, color=color_dict['old_knapsack'])\n",
    "plt.scatter(confidence_thresholds, ratio_new, color=color_dict['new_knapsack'])\n",
    "\n",
    "# create a dash line at y ratio of 20\n",
    "plt.axhline(y=20, color='black', linestyle='--')\n",
    "plt.xlim(-0.1, 1)\n",
    "\n",
    "plt.xlabel('Confidence threshold')\n",
    "plt.ylabel('Mapped to unmapped ratio')\n",
    "plt.legend()\n",
    "plt.title('Mapped to unmapped ratio for different confidence thresholds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_new_knapsack = final[(final[\"confidence_new_knapsack\"] > confidence_threshold_new_knapsack) & (final.protein_id_new_knapsack.notna())]\n",
    "unmatched_new_knapsack = final[(final[\"confidence_new_knapsack\"] > confidence_threshold_new_knapsack) & (final.protein_id_new_knapsack.isna())]\n",
    "\n",
    "matched_new_greedy = final[(final[\"confidence_new_greedy\"] > confidence_threshold_new_greedy) & (final.protein_id_new_greedy.notna())]\n",
    "unmatched_new_greedy = final[(final[\"confidence_new_greedy\"] > confidence_threshold_new_greedy) & (final.protein_id_new_greedy.isna())]\n",
    "\n",
    "matched_old_knapsack = final[(final[\"confidence_old_knapsack\"] > confidence_threshold_old_knapsack) & (final.protein_id.notna())]\n",
    "unmatched_old_knapsack = final[(final[\"confidence_old_knapsack\"] > confidence_threshold_old_knapsack) & (final.protein_id.isna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Due to this higher discrimination of false positives and PSMs in the model confidence scale, IN v1.1 identifies 145.1% more peptides mapping to the proteome at 5% FDR compared to its precursor, leading to a 35.3% increase in protein identifications with the same threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create venn diagram for mapped proteins\n",
    "venn2([get_proteins(matched_new_knapsack, \"protein_id_new_knapsack\"), get_proteins(matched_old_knapsack, \"protein_id\")], set_labels=('IN v1.1 knapsack', 'IN v0.1 knapsack'), set_colors=(color_dict['new_knapsack'], color_dict['old_knapsack']), alpha=0.5)\n",
    "plt.title(f'Mapped proteins between new and old predictions at 5% FDR')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for unique peptide sequences\n",
    "venn2([set(matched_new_knapsack.preds_new_knapsack), set(matched_old_knapsack.preds)], set_labels=('IN v1.1 knapsack', 'IN v0.1 knapsack'), set_colors=(color_dict['new_knapsack'], color_dict['old_knapsack']), alpha=0.5)\n",
    "plt.title('Mapped peptide sequences between new and old predictions at 5% FDR')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein counts for old and new knapsack in the same plot\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.histplot(matched_new_knapsack.groupby('protein_id_new_knapsack').count().preds, discrete=True, color=color_dict['new_knapsack'], label='IN v1.1 knapsack')\n",
    "sns.histplot(matched_old_knapsack.groupby('protein_id').count().preds, discrete=True, color=color_dict['old_knapsack'], label='IN v0 knapsack')\n",
    "plt.title('Distribution of sequences per protein at 5% FDR')\n",
    "plt.xlabel('Number of sequences')\n",
    "plt.ylabel('Number of proteins')\n",
    "plt.legend()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same plot but zoom in the range of 10 to 30\n",
    "plt.figure(figsize=(10, 5))\n",
    "sns.histplot(matched_new_knapsack.groupby('protein_id_new_knapsack').count().preds, discrete=True, color=color_dict['new_knapsack'], label='IN v1.1 knapsack')\n",
    "sns.histplot(matched_old_knapsack.groupby('protein_id').count().preds, discrete=True, color=color_dict['old_knapsack'], label='IN v0 knapsack')\n",
    "plt.title('Distribution of sequences per protein at 5% FDR')\n",
    "plt.xlabel('Number of sequences')\n",
    "plt.ylabel('Number of proteins')\n",
    "plt.legend()\n",
    "plt.xlim(9.5, 26.5)\n",
    "plt.ylim(0, 40)\n",
    "plt.xticks(range(10, 27))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter predictions to mapped and unmapped reads, higher than XX% confidence\n",
    "prediction_confidence_threshold = 0.0\n",
    "\n",
    "matched_new_knapsack = final[(final[\"confidence_new_knapsack\"] > prediction_confidence_threshold) & (final[\"protein_id_new_knapsack\"].notna())]\n",
    "unmatched_new_predictions = final[(final[\"confidence_new_knapsack\"] > prediction_confidence_threshold) & (final[\"protein_id_new_knapsack\"].isna())]\n",
    "\n",
    "print(f\"New Mapped predictions labeled shape: {matched_new_knapsack.shape}\")\n",
    "print(f\"New Unmapped predictions labeled shape: {unmatched_new_knapsack.shape}\")\n",
    "\n",
    "matched_old_knapsack = final[(final[\"confidence_old_knapsack\"] > prediction_confidence_threshold) & (final[\"protein_id\"].notna())]\n",
    "unmatched_old_predictions = final[(final[\"confidence_old_knapsack\"] > prediction_confidence_threshold) & (final[\"protein_id\"].isna())]\n",
    "\n",
    "print(f\"Old Mapped predictions labeled shape: {matched_old_knapsack.shape}\")\n",
    "print(f\"Old Unmapped predictions labeled shape: {unmatched_old_knapsack.shape}\")\n",
    "\n",
    "matched_new_greedy = final[(final[\"confidence_new_greedy\"] > prediction_confidence_threshold) & (final[\"protein_id_new_greedy\"].notna())]\n",
    "unmatched_new_greedy = final[(final[\"confidence_new_greedy\"] > prediction_confidence_threshold) & (final[\"protein_id_new_greedy\"].isna())]\n",
    "\n",
    "print(f\"New Greedy Mapped predictions labeled shape: {matched_new_greedy.shape}\")\n",
    "print(f\"New Greedy Unmapped predictions labeled shape: {unmatched_new_greedy.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create venn diagram for mapped proteins\n",
    "venn2([get_proteins(matched_new_knapsack, \"protein_id_new_knapsack\"), get_proteins(matched_old_knapsack, \"protein_id\")], set_labels=('IN v1.1 knapsack', 'IN v0 knapsack'), set_colors=(color_dict['new_knapsack'], color_dict['old_knapsack']))\n",
    "plt.title('Mapped proteins between new and old predictions, confidence > 0.0')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for unique peptide sequences\n",
    "venn2([set(matched_new_knapsack.preds_new_knapsack), set(matched_old_knapsack.preds)], set_labels=('IN v1.1 knapsack', 'IN v0 knapsack'), set_colors=(color_dict['new_knapsack'], color_dict['old_knapsack']))\n",
    "plt.title('Mapped peptide sequences between new and old predictions, confidence > 0.0')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Expanded modification support, features and runtime\n",
    "\n",
    "The IN v1.1 release introduces compatibility with four additional modifications: phosphorylation, deamidation, carbamylation, and ammonia loss. Carbamidomethylation is now a variable modification, with variable acetylation and oxidation also supported. This results in an increase in overall peptide sequence predictions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count unique number of modifications for old and new knapsack\n",
    "unique_mods_old = set()\n",
    "\n",
    "for seq in final.mod_preds:\n",
    "    if isinstance(seq, str):\n",
    "        res = re.findall(r'\\([a-zA-Z]+\\)', seq)  # Find all modifications in parentheses\n",
    "        unique_mods_old.update(res)  # Add all found modifications to the set\n",
    "\n",
    "# unique_mods_old.add('(carbamidomethyl)')  # Add fixed modification to the set\n",
    "print(unique_mods_old)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for new knapsack\n",
    "unique_mods_new = set()\n",
    "\n",
    "for seq in final.mod_preds_new_knapsack:  # Assuming this is your dataframe column\n",
    "    if isinstance(seq, str):\n",
    "        res = re.findall(r'\\[UNIMOD:\\d+\\]', seq)  # Find all modifications in brackets\n",
    "        unique_mods_new.update(res)  # Add all found modifications to the set\n",
    "\n",
    "print(unique_mods_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = {'[UNIMOD:1]': 'Acetylation', '[UNIMOD:21]': 'Phosphorylation', '[UNIMOD:4]': 'Carbamidomethyl', '[UNIMOD:35]': 'Oxidation', '[UNIMOD:7]': 'Deamidation', '[UNIMOD:385]': 'Ammonia loss', '[UNIMOD:5]': 'Carbamylation'}\n",
    "print(len(mapping))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_mods_new.remove('[UNIMOD:4]')\n",
    "unique_mods_new.remove('[UNIMOD:5]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of unique modifications for old and new knapsack\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v0.1', 'IN v1.1'], y=[len(unique_mods_old), len(unique_mods_new)], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.title('Number of variable modifications')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of peptides with variable modifications for old and new knapsack\n",
    "peptides_with_mods_old = set()\n",
    "for seq in final.mod_preds:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_old:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_old.add(seq)\n",
    "                break\n",
    "\n",
    "len(peptides_with_mods_old)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of peptides with variable modifications for new knapsack\n",
    "peptides_with_mods_new = set()\n",
    "individual_mods_dict = {}\n",
    "for seq in final.mod_preds_new_knapsack:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_new:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_new.add(seq)\n",
    "                break\n",
    "        for mod in unique_mods_new:\n",
    "            if mod in seq:\n",
    "                individual_mods_dict[mod] = individual_mods_dict.get(mod, 0) + 1\n",
    "\n",
    "len(peptides_with_mods_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(individual_mods_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variable modifications for new greedy\n",
    "peptides_with_mods_greedy = set()\n",
    "\n",
    "for seq in final.mod_preds_new_greedy:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_new:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_greedy.add(seq)\n",
    "                break\n",
    "\n",
    "len(peptides_with_mods_greedy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications for old and new knapsack and new greedy\n",
    "plt.figure(figsize=(6, 5))\n",
    "sns.barplot(x=['IN v0.1 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[len(peptides_with_mods_old), len(peptides_with_mods_new), len(peptides_with_mods_greedy)], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Number of peptides with variable modifications')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications for old and new knapsack\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v0', 'IN v1.1'], y=[len(peptides_with_mods_old), len(peptides_with_mods_new)], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.title('Number of peptides with variable modifications, knapsack')\n",
    "plt.tight_layout()\n",
    "sns.despine()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate number of peptides with variable modifications above the 5% FDR threshold for old and new knapsack and new greedy\n",
    "peptides_with_mods_old_fdr = set()\n",
    "for seq in final.mod_preds[final.confidence_old_knapsack >= confidence_threshold_old_knapsack]:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_old:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_old_fdr.add(seq)\n",
    "                break\n",
    "\n",
    "peptides_with_mods_new_fdr = set()\n",
    "for seq in final.mod_preds_new_knapsack[final.confidence_new_knapsack >= confidence_threshold_new_knapsack]:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_new:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_new_fdr.add(seq)\n",
    "                break\n",
    "\n",
    "peptides_with_mods_greedy_fdr = set()\n",
    "for seq in final.mod_preds_new_greedy[final.confidence_new_greedy >= confidence_threshold_new_greedy]:\n",
    "    if isinstance(seq, str):\n",
    "        for mod in unique_mods_new:\n",
    "            if mod in seq:\n",
    "                peptides_with_mods_greedy_fdr.add(seq)\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications above the 5% FDR threshold for old and new knapsack and new greedy\n",
    "plt.figure(figsize=(6, 5))\n",
    "sns.barplot(x=['IN v0.1 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[len(peptides_with_mods_old_fdr), len(peptides_with_mods_new_fdr), len(peptides_with_mods_greedy_fdr)], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Number of peptides with variable modifications at 5% FDR')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see a substantial number of high confidence predictions at 5% FDR, even in samples without enrichment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications above the 5% FDR threshold for old and new knapsack\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v0', 'IN v1.1'], y=[len(peptides_with_mods_old_fdr), len(peptides_with_mods_new_fdr)], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.title('Number of peptides with variable modifications at 5% FDR')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For large datasets, especially when prioritising high-confidence results, there is a tradeoff between performance and inference time. Despite linear scaling, knapsack precursor fit takes 36 times longer than greedy precursor fit per spectrum. While knapsack beam search provides the best performance, it comes at a steep computational cost. In many cases, running the model with greedy precursor mass search may be more efficient while still delivering robust results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot inference times for new knapsack and new greedy\n",
    "inference_time_new_knapsack = 172\n",
    "inference_time_new_greedy = 4.68\n",
    "\n",
    "ratio = inference_time_new_knapsack / inference_time_new_greedy\n",
    "\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v1.1 knapsack', 'IN v1.1 greedy'], y=[inference_time_new_knapsack, inference_time_new_greedy], palette=[color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.ylabel('Inference time (ms)')\n",
    "plt.title('Inference time for new knapsack and new greedy')\n",
    "plt.xlabel(f'Greedy search is {ratio:.2f} times faster')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_mods_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate column with mod or no mod\n",
    "final['mod_knapsack_old'] = np.nan\n",
    "final['mod_knapsack_new'] = np.nan\n",
    "final['mod_greedy'] = np.nan\n",
    "\n",
    "for i, row in final.iterrows():\n",
    "    for mod in unique_mods_old:\n",
    "        if isinstance(row['mod_preds'], str):\n",
    "            if mod in row['mod_preds']:\n",
    "                final.at[i, 'mod_knapsack_old'] = 'mod'\n",
    "                break\n",
    "            else:\n",
    "                final.at[i, 'mod_knapsack_old'] = 'no mod'\n",
    "    for mod in unique_mods_new:\n",
    "        if isinstance(row['mod_preds_new_knapsack'], str):\n",
    "            if mod in row['mod_preds_new_knapsack']:\n",
    "                final.at[i, 'mod_knapsack_new'] = 'mod'\n",
    "                break\n",
    "            else:\n",
    "                final.at[i, 'mod_knapsack_new'] = 'no mod'\n",
    "    for mod in unique_mods_new:\n",
    "        if isinstance(row['mod_preds_new_greedy'], str):\n",
    "            if mod in row['mod_preds_new_greedy']:\n",
    "                final.at[i, 'mod_greedy'] = 'mod'\n",
    "                break\n",
    "            else:\n",
    "                final.at[i, 'mod_greedy'] = 'no mod'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find number of peptides with variable modifications for old and new knapsack and new greedy that map to the proteome\n",
    "psms_old_knapsack_mod = final[(final['mod_knapsack_old'] == 'mod') & (final['preds'] == final['stripped_target'])].shape[0]\n",
    "psms_new_knapsack_mod = final[(final['mod_knapsack_new'] == 'mod') & (final['protein_id_new_knapsack'].notna())].shape[0]\n",
    "psms_new_greedy_mod = final[(final['mod_greedy'] == 'mod') & (final['protein_id_new_greedy'].notna())].shape[0]\n",
    "\n",
    "print(f'Old Knapsack: {psms_old_knapsack_mod} PSMs with variable modifications')\n",
    "print(f'New Knapsack: {psms_new_knapsack_mod} PSMs with variable modifications')\n",
    "print(f'New Greedy: {psms_new_greedy_mod} PSMs with variable modifications')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications for old and new knapsack and new greedy that map to the proteome\n",
    "plt.figure(figsize=(6, 5))\n",
    "sns.barplot(x=['IN v0 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[psms_old_knapsack_mod, psms_new_knapsack_mod, psms_new_greedy_mod], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Number of PSMs with variable modifications that map to the proteome')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for 5% FDR\n",
    "psms_old_knapsack_mod_fdr = final[(final['mod_knapsack_old'] == 'mod') & (final['preds'] == final['stripped_target']) & (final['confidence_old_knapsack'] >= confidence_threshold_old_knapsack)].shape[0]\n",
    "psms_new_knapsack_mod_fdr = final[(final['mod_knapsack_new'] == 'mod') & (final['protein_id_new_knapsack'].notna()) & (final['confidence_new_knapsack'] >= confidence_threshold_new_knapsack)].shape[0]\n",
    "psms_new_greedy_mod_fdr = final[(final['mod_greedy'] == 'mod') & (final['protein_id_new_greedy'].notna()) & (final['confidence_new_greedy'] >= confidence_threshold_new_greedy)].shape[0]\n",
    "\n",
    "print(f'Old Knapsack: {psms_old_knapsack_mod_fdr} PSMs with variable modifications at 5% FDR')\n",
    "print(f'New Knapsack: {psms_new_knapsack_mod_fdr} PSMs with variable modifications at 5% FDR')\n",
    "print(f'New Greedy: {psms_new_greedy_mod_fdr} PSMs with variable modifications at 5% FDR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of peptides with variable modifications for old and new knapsack and new greedy that map to the proteome at 5% FDR\n",
    "plt.figure(figsize=(6, 5))\n",
    "sns.barplot(x=['IN v0 knapsack', 'IN v1.1 knapsack', 'IN v1.1 greedy'], y=[psms_old_knapsack_mod_fdr, psms_new_knapsack_mod_fdr, psms_new_greedy_mod_fdr], palette=[color_dict['old_knapsack'], color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Number of PSMs with variable modifications that map to the proteome at 5% FDR')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same without greedy\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v0', 'IN v1.1'], y=[psms_old_knapsack_mod_fdr, psms_new_knapsack_mod_fdr], palette=[color_dict['old_knapsack'], color_dict['new_knapsack']])\n",
    "plt.title('Number of PSMs with variable modifications that map to the proteome at 5% FDR')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the updated version, a spectrum dataframe class has been introduced to further improve data import robustness and interoperability. Residue log probabilities are now included in the results alongside the token list, and scan number and precursor mass error (in ppm) are now also included in the output table. The model also supports evaluation mode with a database search reference, where FDR confidence thresholds are automatically reported during inference. In the new model, leucine and isoleucine are predicted as separate residue tokens, albeit with relatively low recall at peptide level (17.7% for leucine and 19.6% for isoleucine)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# evaluate recall in PSMs with isoleucine and leucine\n",
    "# get all PSMs with isoleucine and leucine\n",
    "psms_with_il_new_knapsack = final[(final['preds_new_knapsack'].str.contains('I')) | (final['preds_new_knapsack'].str.contains('L'))].shape[0]\n",
    "psms_with_il_new_greedy = final[(final['preds_new_greedy'].str.contains('I')) | (final['preds_new_greedy'].str.contains('L'))].shape[0]\n",
    "\n",
    "# get all PSMs with isoleucine and leucine that have targets\n",
    "psms_with_i_target_new_knapsack = final[(final['preds_new_knapsack'].str.contains('I')) & (final['sequence'].notna())].shape[0]\n",
    "psms_with_i_target_new_greedy = final[(final['preds_new_greedy'].str.contains('I')) & (final['sequence'].notna())].shape[0]\n",
    "\n",
    "print(f'New knapsack: {psms_with_i_target_new_knapsack} PSMs with isoleucine and leucine')\n",
    "print(f'New greedy: {psms_with_i_target_new_greedy} PSMs with isoleucine and leucine')\n",
    "\n",
    "# get all PSMs with isoleucine and leucine that have targets and are correct\n",
    "psms_with_i_target_correct_new_knapsack = final[(final['preds_new_knapsack'].str.contains('I')) & (final['sequence'].notna()) & (final['preds_new_knapsack'] == final['sequence'])].shape[0]\n",
    "psms_with_i_target_correct_new_greedy = final[(final['preds_new_greedy'].str.contains('I')) & (final['sequence'].notna()) & (final['preds_new_greedy'] == final['sequence'])].shape[0]\n",
    "\n",
    "print(f'New knapsack: {psms_with_i_target_correct_new_knapsack} PSMs with isoleucine and leucine that are correct')\n",
    "print(f'New greedy: {psms_with_i_target_correct_new_greedy} PSMs with isoleucine and leucine that are correct')\n",
    "\n",
    "# calculate recall\n",
    "recall_new_knapsack = psms_with_i_target_correct_new_knapsack / psms_with_i_target_new_knapsack\n",
    "recall_new_greedy = psms_with_i_target_correct_new_greedy / psms_with_i_target_new_greedy\n",
    "\n",
    "print(f'Recall for IN v1.1 knapsack: {recall_new_knapsack:.3f}')\n",
    "print(f'Recall for IN v1.1 greedy: {recall_new_greedy:.3f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot both recalls for both amino acids for new knapsack and new greedy\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v1.1 knapsack', 'IN v1.1 greedy'], y=[recall_new_knapsack, recall_new_greedy], palette=[color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Peptide recall for isoleucine')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for leucine\n",
    "psms_with_l_target_new_knapsack = final[(final['preds_new_knapsack'].str.contains('L')) & (final['sequence'].notna())].shape[0]\n",
    "psms_with_l_target_new_greedy = final[(final['preds_new_greedy'].str.contains('L')) & (final['sequence'].notna())].shape[0]\n",
    "\n",
    "psms_with_l_target_correct_new_knapsack = final[(final['preds_new_knapsack'].str.contains('L')) & (final['sequence'].notna()) & (final['preds_new_knapsack'] == final['sequence'])].shape[0]\n",
    "psms_with_l_target_correct_new_greedy = final[(final['preds_new_greedy'].str.contains('L')) & (final['sequence'].notna()) & (final['preds_new_greedy'] == final['sequence'])].shape[0]\n",
    "\n",
    "recall_new_knapsack = psms_with_l_target_correct_new_knapsack / psms_with_l_target_new_knapsack\n",
    "recall_new_greedy = psms_with_l_target_correct_new_greedy / psms_with_l_target_new_greedy\n",
    "\n",
    "print(f'Recall for IN v1.1 knapsack: {recall_new_knapsack:.3f}')\n",
    "print(f'Recall for IN v1.1 greedy: {recall_new_greedy:.3f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot both recalls for both amino acids for new knapsack and new greedy\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.barplot(x=['IN v1.1 knapsack', 'IN v1.1 greedy'], y=[recall_new_knapsack, recall_new_greedy], palette=[color_dict['new_knapsack'], color_dict['new_greedy']])\n",
    "plt.title('Peptide recall for leucine')\n",
    "plt.tight_layout()\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although here we focus on a single dataset, we have observed similar performance gains across biological samples and applications, indicating generalized gains. Importantly, we observe an impressive 81.3% peptide recall in our GluC dataset, which is a GluC pretreated HeLa proteome with more data than the HeLa QC above.\n",
    "\n",
    "We believe our models continue to improve in performance, and we have yet to determine the upper limit for database search recall and novel identification rates. Our ultimate goal is to achieve robust accuracy near the 90% recall mark across experimental datasets, edging closer to solving de novo peptide sequencing. We will continue to scale and refine our models, as well as explore biological applications and features further. Additionally, the next generation of IN+, our diffusion model, is being trained and will be soon released in a similar post. \n",
    "\n",
    "So stay tuned for more updates, several preprints coming out soon! \n",
    "\n",
    "Original research paper available at: [InstaNovo enables diffusion-powered de novo peptide sequencing in large-scale proteomics experiments](https://www.nature.com/articles/s42256-025-01019-5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
