{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0",
   "metadata": {
    "id": "2f77bd43-f74e-4e80-b51a-b9e619d2ba11",
    "tags": []
   },
   "source": [
    "# Getting started with InstaNovo\n",
    "\n",
    "<a target=\"_blank\" href=\"https://colab.research.google.com/github/instadeepai/InstaNovo/blob/main/notebooks/getting_started_with_instanovo.ipynb\">\n",
    "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>\n",
    "</a>\n",
    "<!-- <a target=\"_blank\" href=\"https://kaggle.com/kernels/welcome?src=https://github.com/instadeepai/InstaNovo/blob/main/notebooks/getting_started_with_instanovo.ipynb\">\n",
    "<img src=\"https://kaggle.com/static/images/open-in-kaggle.svg\" alt=\"Open In Kaggle\"/> </a> -->\n",
    "\n",
    "In this notebook, we demo InstaNovo, a transformer neural network with the ability to translate fragment ion peaks into the sequence of amino acids that make up the studied peptide(s). We evaluate the model on the yeast test fold of nine-species dataset."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1",
   "metadata": {
    "id": "v4Rk9kj1NiMU"
   },
   "source": [
    "![](https://raw.githubusercontent.com/instadeepai/InstaNovo/main/docs/assets/graphical_abstract.jpeg)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2",
   "metadata": {
    "id": "VwB6HeU1Ngxk"
   },
   "source": [
    "**Links:**\n",
    "\n",
    "- Nature Machine Intelligence Paper: [**InstaNovo enables diffusion-powered de novo peptide sequencing in large-scale proteomics experiments**](https://www.nature.com/articles/s42256-025-01019-5) \\\n",
    "  Kevin Eloff, Konstantinos Kalogeropoulos, Amandla Mabona, Oliver Morell, Rachel Catzel, Esperanza Rivera-de-Torre, Jakob Berg Jespersen, Wesley Williams, Sam P. B. van Beljouw, Marcin J. Skwark, Andreas Hougaard Laustsen, Stan J. J. Brouns, Anne Ljungars, Erwin M. Schoof, Jeroen Van Goey, Ulrich auf dem Keller, Karim Beguir, Nicolas Lopez Carranza, Timothy P. Jenkins\n",
    "- Code: [GitHub](https://github.com/instadeepai/InstaNovo)\n",
    "\n",
    "**Important:**\n",
    "\n",
    "It is highly recommended to run this notebook in an environment with access to a GPU. If you are running this notebook in Google Colab:\n",
    "\n",
    "- In the menu, go to `Runtime > Change Runtime Type > T4 GPU`"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3",
   "metadata": {
    "id": "be4f1d34-df21-453f-b167-0ef935c85f00"
   },
   "source": [
    "## Loading the InstaNovo model\n",
    "\n",
    "We first install the latest instanovo from PyPi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "if 'google.colab' in sys.modules or 'KAGGLE_KERNEL_RUN_TYPE' in os.environ:\n",
    "    # Suppress TensorFlow warnings\n",
    "    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'\n",
    "    # Install torchvision when running on Google Colab to prevent errors\n",
    "    !uv pip install --system \"instanovo[cu124]>=1.1.1\" pyopenms-viz torchvision tf-nightly\n",
    "else:\n",
    "    !uv pip install \"instanovo[cu124]>=1.1.1\" pyopenms-viz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter warnings and set logging level\n",
    "import warnings\n",
    "import logging\n",
    "warnings.filterwarnings(\"ignore\", module=\"matplotlib\")\n",
    "warnings.filterwarnings(\"ignore\", module=\"torch\")\n",
    "logging.getLogger(\"matplotlib\").setLevel(logging.WARNING)\n",
    "logging.getLogger(\"rdkit\").setLevel(logging.WARNING)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "We can use `instanovo version` to check the version of InstaNovo (the transformer-based model), InstaNovo+ (the diffusion-based model) and some of their dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!instanovo version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8",
   "metadata": {},
   "source": [
    "Import the transformer-based InstaNovo model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "cc618a03-d365-484d-a001-9b991141d1f5",
    "outputId": "4676a7de-cb22-4cc4-8811-b6e9da92172a",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from instanovo.transformer.model import InstaNovo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10",
   "metadata": {},
   "source": [
    "Set the device to GPU if available (recommended), otherwise use CPU."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 35
    },
    "id": "d8c02b31-e704-49cd-9dad-60a2f5b5cfa3",
    "outputId": "edce16ce-b370-4430-da76-07b41a1de3ff",
    "tags": []
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "device"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "12",
   "metadata": {
    "id": "7QcyM4jKA9qL"
   },
   "source": [
    "InstaNovo supports automatic model downloads. You can see the IDs of the pretrained models that are available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "InstaNovo.get_pretrained()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14",
   "metadata": {},
   "source": [
    "And download the model checkpoint given the ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {
    "id": "32eaf403-a1b8-4737-96ed-4931c5db44e6",
    "tags": []
   },
   "outputs": [],
   "source": [
    "model, config = InstaNovo.from_pretrained(\"instanovo-v1.1.0\")\n",
    "model = model.to(device).eval()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16",
   "metadata": {},
   "source": [
    "Alternatively, you can also download the model checkpoint directly from the [InstaNovo releases](https://github.com/instadeepai/InstaNovo/releases) page."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "17",
   "metadata": {
    "id": "bcdeeddb-6b91-4cd1-bb30-f72d97d809de"
   },
   "source": [
    "## Loading the nine-species dataset\n",
    "Download the [yeast test fold of the nine-species dataset](https://huggingface.co/datasets/InstaDeepAI/instanovo_ninespecies_exclude_yeast) dataset from HuggingFace.\n",
    "\n",
    "We can use our SpectrumDataFrame to download this directly. SpectrumDataFrame is a special data class used by InstaNovo to read and write from multiple formats, including mgf, mzml, mzxml, pandas, polars, HuggingFace, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "from instanovo.utils import SpectrumDataFrame\n",
    "\n",
    "sdf = SpectrumDataFrame.from_huggingface(\n",
    "    \"InstaDeepAI/ms_ninespecies_benchmark\",\n",
    "    is_annotated=True,\n",
    "    shuffle=False,\n",
    "    split=\"test[:10%]\",  # Let's only use a subset of the test data for faster inference in this notebook\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdf.to_pandas().head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20",
   "metadata": {},
   "source": [
    "Let's quickly plot the spectrum in the first row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pd.options.plotting.backend = \"ms_matplotlib\"\n",
    "row = sdf[0]\n",
    "row_df = pd.DataFrame({\"mz\": row[\"mz_array\"], \"intensity\": row[\"intensity_array\"]})\n",
    "row_df.plot(\n",
    "    kind=\"spectrum\",\n",
    "    x=\"mz\",\n",
    "    y=\"intensity\",\n",
    "    annotate_mz=True,\n",
    "    bin_method=\"none\",\n",
    "    annotate_top_n_peaks=5,\n",
    "    aggregate_duplicates=True,\n",
    "    title=f\"Mass spectrum of {row['sequence']}\",\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {
    "id": "a088e7c3-93ba-4421-879a-a5fbb3c90a7f",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from instanovo.transformer.dataset import SpectrumDataset, collate_batch\n",
    "\n",
    "ds = SpectrumDataset(\n",
    "    sdf,\n",
    "    model.residue_set,\n",
    "    config.get(\"n_peaks\", 200),\n",
    "    return_str=True,\n",
    "    annotated=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {
    "id": "ac42a78b-942b-4996-a5e7-e4d888091fda",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from torch.utils.data import DataLoader\n",
    "\n",
    "# When using SpectrumDataFrame, workers and shuffle is handled internally.\n",
    "dl = DataLoader(ds, batch_size=64, shuffle=False, num_workers=0, collate_fn=collate_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {
    "id": "8e51fdeb-ae0f-4851-802f-d692504ee349",
    "tags": []
   },
   "outputs": [],
   "source": [
    "batch = next(iter(dl))\n",
    "\n",
    "spectra, precursors, spectra_mask, peptides, _ = batch\n",
    "spectra = spectra.to(device)\n",
    "precursors = precursors.to(device)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "25",
   "metadata": {
    "id": "8a16c311-6802-49f8-af8e-857f43510c37"
   },
   "source": [
    "## Decoding\n",
    "\n",
    "We have three options for decoding:\n",
    "- Greedy Search\n",
    "- Beam Search\n",
    "- Knapsack Beam Search\n",
    "\n",
    "For the best results and highest peptide recall, use **Knapsack Beam Search**. \n",
    "For fastest results (over 10x speedup), use **Greedy Search**.\n",
    "\n",
    "We generally use a beam size of 5 for Beam Search and Knapsack Beam Search, a higher beam size should increase recall at the cost of performance and vice versa.\n",
    "\n",
    "_Note: in our findings, greedy search has similar performance as knapsack beam search at 5% FDR. I.e. if you plan to filter at 5% FDR anyway, use greedy search for optimal performance._\n",
    "\n",
    "### Greedy Search and Beam Search\n",
    "\n",
    "Greedy search is used when `num_beams=1`, and beam search is used when `num_beams>1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "from instanovo.inference import BeamSearchDecoder, GreedyDecoder\n",
    "\n",
    "num_beams = 1  # Change this, defaults are 1 or 5\n",
    "\n",
    "if num_beams > 1:\n",
    "    decoder = BeamSearchDecoder(model=model)\n",
    "else:\n",
    "    decoder = GreedyDecoder(model=model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27",
   "metadata": {},
   "source": [
    "### Knapsack Beam Search\n",
    "\n",
    "Setup knapsack beam search decoder. This may take a few minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3537de13-8cbb-4893-86cf-f33fbd7ee567",
    "outputId": "5c6cad7f-8b11-4cbe-da00-de08850c1712",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from instanovo.constants import MASS_SCALE\n",
    "from instanovo.inference.knapsack import Knapsack\n",
    "from instanovo.inference.knapsack_beam_search import KnapsackBeamSearchDecoder\n",
    "\n",
    "num_beams = 5\n",
    "\n",
    "def _setup_knapsack(model: InstaNovo) -> Knapsack:\n",
    "    # Cannot allow negative masses in knapsack graph\n",
    "    if \"(-17.03)\" in model.residue_set.residue_masses:\n",
    "        model.residue_set.residue_masses[\"(-17.03)\"] = 1e3\n",
    "    if \"[UNIMOD:385]\" in model.residue_set.residue_masses:\n",
    "        model.residue_set.residue_masses[\"[UNIMOD:385]\"] = 1e3\n",
    "\n",
    "    residue_masses = dict(model.residue_set.residue_masses.copy())\n",
    "    if any(x < 0 for x in residue_masses.values()):\n",
    "        raise NotImplementedError(\n",
    "            \"Negative mass found in residues, this will break the knapsack graph. \"\n",
    "            \"Either disable knapsack or use strictly positive masses\"\n",
    "        )\n",
    "    for special_residue in list(model.residue_set.residue_to_index.keys())[:3]:\n",
    "        residue_masses[special_residue] = 0\n",
    "    residue_indices = model.residue_set.residue_to_index\n",
    "    return Knapsack.construct_knapsack(\n",
    "        residue_masses=residue_masses,\n",
    "        residue_indices=residue_indices,\n",
    "        max_mass=4000.00,\n",
    "        mass_scale=MASS_SCALE,\n",
    "    )\n",
    "\n",
    "\n",
    "knapsack_path = Path(\"./checkpoints/knapsack/\")\n",
    "\n",
    "if not knapsack_path.exists():\n",
    "    print(\"Knapsack path missing or not specified, generating...\")\n",
    "    knapsack = _setup_knapsack(model)\n",
    "    decoder = KnapsackBeamSearchDecoder(model, knapsack)\n",
    "    print(f\"Saving knapsack to {knapsack_path}\")\n",
    "    knapsack_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "    knapsack.save(knapsack_path)\n",
    "else:\n",
    "    print(\"Knapsack path found. Loading...\")\n",
    "    decoder = KnapsackBeamSearchDecoder.from_file(model=model, path=knapsack_path)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "29",
   "metadata": {
    "id": "vBZwQvIYB2Wa"
   },
   "source": [
    "## Inference time ðŸš€"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "30",
   "metadata": {
    "id": "0ad8db32-98c4-4ba2-a58a-719d194345ec"
   },
   "source": [
    "Evaluating a single batch..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {
    "id": "52ff9613-9670-48ae-afd7-78f78efa71d0",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from instanovo.inference import ScoredSequence\n",
    "\n",
    "with torch.no_grad():\n",
    "    p = decoder.decode(\n",
    "        spectra=spectra,\n",
    "        precursors=precursors,\n",
    "        beam_size=num_beams,\n",
    "        max_length=config[\"max_length\"],\n",
    "    )\n",
    "\n",
    "preds = [x.sequence if isinstance(x, ScoredSequence) else [] for x in p]\n",
    "probs = [x.sequence_log_probability if isinstance(x, ScoredSequence) else -float(\"inf\") for x in p]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32",
   "metadata": {},
   "source": [
    "### Confidence probabilities\n",
    "The model returns per-residue confidences in the form of token log-probabilities. We can visualize these or use them as part of a workflow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Optional\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "from instanovo.inference.beam_search import ScoredSequence\n",
    "\n",
    "\n",
    "def plot_residue_confidence(prediction: ScoredSequence, peptide: Optional[str] = None) -> None:\n",
    "    if not prediction:\n",
    "        return\n",
    "    ticks = list(range(len(prediction.sequence)))\n",
    "    token_probabilities = np.exp(prediction.token_log_probabilities[:len(ticks)])\n",
    "    sequence_confidence = np.exp(prediction.sequence_log_probability)\n",
    "\n",
    "    _, ax = plt.subplots()\n",
    "    bars = sns.barplot(x=ticks, y=token_probabilities, errorbar=None, ax=ax)\n",
    "\n",
    "    # Increase Y-axis limit to create space for text labels\n",
    "    ax.set_ylim(0, max(token_probabilities) * 1.2)\n",
    "\n",
    "    # Add numbers above bars with a slanted angle\n",
    "    for bar, prob in zip(bars.patches, token_probabilities):\n",
    "        height = bar.get_height()\n",
    "        ax.text(\n",
    "            bar.get_x() + bar.get_width() / 2,\n",
    "            float(height) + 0.02,\n",
    "            f\"{float(prob):.4f}\",\n",
    "            ha=\"center\",\n",
    "            va=\"bottom\",\n",
    "            fontsize=9,\n",
    "            color=\"black\",\n",
    "            rotation=45,\n",
    "        )\n",
    "\n",
    "    # Check if any residue contains a PTM (e.g., \"S(+79.97)\")\n",
    "    has_ptm = any(\"(\" in res and \")\" in res for res in prediction.sequence)\n",
    "\n",
    "    # Set labels\n",
    "    x_label = f\"    Prediction: {''.join(prediction.sequence)}\"\n",
    "    if peptide is not None:\n",
    "        x_label += f\"\\nGround truth: {peptide}\"\n",
    "    ax.set_xlabel(x_label)\n",
    "    ax.set_ylabel(\"Confidence Probability\")\n",
    "\n",
    "    # Set title with sequence confidence\n",
    "    ax.set_title(\n",
    "        f\"Residue Confidence per Position\\nSequence Probability: {sequence_confidence:.4f}\"\n",
    "    )\n",
    "\n",
    "    # Set X-ticks\n",
    "    ax.set_xticks(ticks)\n",
    "    ax.set_xticklabels(\n",
    "        prediction.sequence,\n",
    "        rotation=45 if has_ptm else 0,\n",
    "        ha=\"right\" if has_ptm else \"center\",\n",
    "    )\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34",
   "metadata": {},
   "source": [
    "For a spectrum that is sequenced correctly, the sequence probability and per-residue probabilities are uniformly high:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_residue_confidence(p[-1], peptides[-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36",
   "metadata": {},
   "source": [
    "For another spectrum which is sequenced incorrectly, the sequence probability is low and the per-residue probabilities of the incorrectly sequenced residues (up to isomerism) are lower than of those correctly sequenced:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_residue_confidence(p[0], peptides[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38",
   "metadata": {},
   "source": [
    "These examples suggest the model is fairly well calibrated."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39",
   "metadata": {},
   "source": [
    "### Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {
    "id": "d61eebdf-abb8-449d-b7dd-dce6273747a0",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from instanovo.utils.metrics import Metrics\n",
    "\n",
    "metrics = Metrics(model.residue_set, config[\"isotope_error_range\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41",
   "metadata": {},
   "source": [
    "We include a residue remapping to ensure our input dataset can be mapped to the format the model vocabulary expects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.residue_set.update_remapping({\n",
    "  \"M(ox)\": \"M[UNIMOD:35]\",\n",
    "  \"M(+15.99)\": \"M[UNIMOD:35]\",\n",
    "  \"S(p)\": \"S[UNIMOD:21]\", # Phosphorylation\n",
    "  \"T(p)\": \"T[UNIMOD:21]\",\n",
    "  \"Y(p)\": \"Y[UNIMOD:21]\",\n",
    "  \"S(+79.97)\": \"S[UNIMOD:21]\",\n",
    "  \"T(+79.97)\": \"T[UNIMOD:21]\",\n",
    "  \"Y(+79.97)\": \"Y[UNIMOD:21]\",\n",
    "  \"Q(+0.98)\": \"Q[UNIMOD:7]\", # Deamidation\n",
    "  \"N(+0.98)\": \"N[UNIMOD:7]\",\n",
    "  \"Q(+.98)\": \"Q[UNIMOD:7]\",\n",
    "  \"N(+.98)\": \"N[UNIMOD:7]\",\n",
    "  \"C(+57.02)\": \"C[UNIMOD:4]\", # Carboxyamidomethylation\n",
    "  \"(+42.01)\": \"[UNIMOD:1]\", # Acetylation\n",
    "  \"(+43.01)\": \"[UNIMOD:5]\", # Carbamylation\n",
    "  \"(-17.03)\": \"[UNIMOD:385]\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "92270452-2371-484f-8cdd-c1f6492aa4bf",
    "outputId": "6b9c0a83-7a41-4815-9c0e-ea891272ec69",
    "tags": []
   },
   "outputs": [],
   "source": [
    "aa_precision, aa_recall, peptide_recall, peptide_precision = metrics.compute_precision_recall(\n",
    "    peptides, preds\n",
    ")\n",
    "peptide_recall"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "44",
   "metadata": {
    "id": "b6d3e928-2b46-45a0-bdd3-d86970ba4ece"
   },
   "source": [
    "Evaluating on the yeast test fold of the nine-species dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "f0171e22-b893-4876-b1da-4603ad022136",
    "outputId": "9d7233ed-c579-4d57-954a-fafc52f43218",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from tqdm.notebook import tqdm\n",
    "\n",
    "preds = []\n",
    "targs = []\n",
    "probs = []\n",
    "\n",
    "for _, batch in tqdm(enumerate(dl), total=len(dl)):\n",
    "    spectra, precursors, _, peptides, _ = batch\n",
    "    spectra = spectra.to(device)\n",
    "    precursors = precursors.to(device)\n",
    "\n",
    "    with torch.no_grad():\n",
    "        p = decoder.decode(\n",
    "            spectra=spectra,\n",
    "            precursors=precursors,\n",
    "            beam_size=config[\"n_beams\"],\n",
    "            max_length=config[\"max_length\"],\n",
    "        )\n",
    "\n",
    "    preds += [x.sequence if isinstance(x, ScoredSequence) else [] for x in p]\n",
    "    probs += [\n",
    "        x.sequence_log_probability if isinstance(x, ScoredSequence) else -float(\"inf\") for x in p\n",
    "    ]\n",
    "    targs += list(peptides)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46",
   "metadata": {},
   "source": [
    "### Evaluation metrics\n",
    "\n",
    "Model performance without filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_precision, aa_recall, peptide_recall, peptide_precision = metrics.compute_precision_recall(\n",
    "    targs, preds\n",
    ")\n",
    "aa_error_rate = metrics.compute_aa_er(targs, preds)\n",
    "auc = metrics.calc_auc(targs, preds, np.exp(pd.Series(probs)))\n",
    "\n",
    "print(f\"amino acid error rate:    {aa_error_rate:.5f}\")\n",
    "print(f\"amino acid precision:     {aa_precision:.5f}\")\n",
    "print(f\"amino acid recall:        {aa_recall:.5f}\")\n",
    "print(f\"peptide precision:        {peptide_precision:.5f}\")\n",
    "print(f\"peptide recall:           {peptide_recall:.5f}\")\n",
    "print(f\"area under the PR curve:  {auc:.5f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48",
   "metadata": {},
   "source": [
    "### We can find a threshold to ensure a desired FDR:\n",
    "\n",
    "Model performance at 5% FDR:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49",
   "metadata": {},
   "outputs": [],
   "source": [
    "fdr = 5 / 100  # Desired FDR\n",
    "\n",
    "_, threshold = metrics.find_recall_at_fdr(targs, preds, np.exp(probs), fdr=fdr)\n",
    "aa_precision_fdr, aa_recall_fdr, peptide_recall_fdr, peptide_precision_fdr = (\n",
    "    metrics.compute_precision_recall(targs, preds, np.exp(probs), threshold=threshold)\n",
    ")\n",
    "print(f\"Performance at {fdr*100:.1f}% FDR:\\n\")\n",
    "print(f\"amino acid precision:     {aa_precision_fdr:.5f}\")\n",
    "print(f\"amino acid recall:        {aa_recall_fdr:.5f}\")\n",
    "print(f\"peptide precision:        {peptide_precision_fdr:.5f}\")\n",
    "print(f\"peptide recall:           {peptide_recall_fdr:.5f}\")\n",
    "print(f\"area under the PR curve:  {auc:.5f}\")\n",
    "print(f\"confidence threshold:     {threshold:.5f}  <-- Use this as a confidence cutoff\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "50",
   "metadata": {
    "id": "IcstKaUGB8Bo"
   },
   "source": [
    "_Note: to reproduce the results of the paper, the entire Yeast test set should be evaluated with the 0.1.7 release of InstaNovo._"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "51",
   "metadata": {
    "id": "ychXR1M3CbKf"
   },
   "source": [
    "### Saving the predictions..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "u4DxD6ZDCMGC",
    "outputId": "d85757be-2ee0-41a8-cf95-b0a7fd0dc015"
   },
   "outputs": [],
   "source": [
    "pred_df = pd.DataFrame(\n",
    "    {\n",
    "        \"targets\": targs,\n",
    "        \"tokenized_predictions\": preds,\n",
    "        \"predictions\": [\"\".join(x) for x in preds],\n",
    "        \"log_probabilities\": probs,\n",
    "    }\n",
    ")\n",
    "pred_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53",
   "metadata": {
    "id": "HJgMPD0YCWIm"
   },
   "outputs": [],
   "source": [
    "pred_df.to_csv(\"predictions_knapsack_beam_search.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54",
   "metadata": {},
   "source": [
    "## InstaNovo+: Iterative Refinement with a Diffusion Model\n",
    "\n",
    "In this section, we show how to refine the predictions from the transformer model with a diffusion model.\n",
    "\n",
    "First, we download the model checkpoint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from instanovo.diffusion.multinomial_diffusion import InstaNovoPlus\n",
    "\n",
    "InstaNovoPlus.get_pretrained()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_model, diffusion_config = InstaNovoPlus.from_pretrained(\"instanovoplus-v1.1.0-alpha\")\n",
    "diffusion_model = diffusion_model.to(device).eval()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57",
   "metadata": {},
   "source": [
    "Next we create a decoder object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58",
   "metadata": {},
   "outputs": [],
   "source": [
    "from instanovo.inference.diffusion import DiffusionDecoder\n",
    "\n",
    "diffusion_decoder = DiffusionDecoder(model=diffusion_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59",
   "metadata": {},
   "source": [
    "Then we prepare the inference data loader using predictions from the InstaNovo transformer model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "\n",
    "diffusion_df = sdf.to_polars(return_lazy=False)\n",
    "\n",
    "diffusion_df = diffusion_df.with_columns(pl.Series(\"sequence\", pred_df[\"tokenized_predictions\"]))\n",
    "\n",
    "diffusion_df = diffusion_df.with_columns(pl.Series(\"original_peptide\", targs)) # preserve targets\n",
    "\n",
    "diffusion_df = diffusion_df.filter(pl.col(\"sequence\") != []) # remove rows with empty predictions\n",
    "targs = diffusion_df[\"original_peptide\"].to_list() # update targets accordingly\n",
    "\n",
    "diffusion_sdf = SpectrumDataFrame.from_polars(df=diffusion_df, is_annotated=False, shuffle=False)\n",
    "\n",
    "diffusion_ds = SpectrumDataset(\n",
    "    diffusion_sdf,\n",
    "    diffusion_model.residues,\n",
    "    diffusion_model.config.get(\"n_peaks\", 200),\n",
    "    return_str=False,\n",
    "    annotated=True,\n",
    "    peptide_pad_length=diffusion_model.config.get(\"max_length\", 30),\n",
    "    diffusion=True,\n",
    "    reverse_peptide=False,  # we do not reverse peptide for diffusion\n",
    "    tokenize_peptide=False,\n",
    ")\n",
    "\n",
    "diffusion_dl = DataLoader(\n",
    "    diffusion_ds,\n",
    "    batch_size=64,\n",
    "    num_workers=0,  # sdf requirement, handled internally\n",
    "    shuffle=False,  # sdf requirement, handled internally\n",
    "    collate_fn=collate_batch,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61",
   "metadata": {},
   "source": [
    "Finally, we predict sequences by iterating over the spectra and refining the InstaNovo predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = []\n",
    "log_probs = []\n",
    "\n",
    "for batch in tqdm(diffusion_dl, total=len(diffusion_dl)):\n",
    "    spectra, precursors, spectra_padding_mask, peptides, _ = batch\n",
    "    spectra = spectra.to(device)\n",
    "    precursors = precursors.to(device)\n",
    "    spectra_padding_mask = spectra_padding_mask.to(device)\n",
    "    peptides = peptides.to(device)\n",
    "    with torch.no_grad():\n",
    "        batch_predictions, batch_log_probs = diffusion_decoder.decode(\n",
    "            spectra=spectra,\n",
    "            spectra_padding_mask=spectra_padding_mask,\n",
    "            precursors=precursors,\n",
    "            initial_sequence=peptides,\n",
    "        )\n",
    "    predictions.extend(batch_predictions)\n",
    "    log_probs.extend(batch_log_probs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63",
   "metadata": {},
   "source": [
    "Iterative refinement improves performance on this sample of the Nine Species dataset. (To replicate the performance reported in the paper, you would need to evaluate on the entire dataset.) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    aa_precision_refined,\n",
    "    aa_recall_refined,\n",
    "    peptide_recall_refined,\n",
    "    peptide_precision_refined,\n",
    ") = metrics.compute_precision_recall(targs, predictions=predictions)\n",
    "aa_error_rate_refined = metrics.compute_aa_er(targs, predictions)\n",
    "auc_refined = metrics.calc_auc(targs, predictions, np.exp(pd.Series(log_probs)))\n",
    "\n",
    "print(f\"amino acid error rate:     {aa_error_rate_refined:.5f}\")\n",
    "print(f\"amino acid precision:      {aa_precision_refined:.5f}\")\n",
    "print(f\"amino acid recall:         {aa_recall_refined:.5f}\")\n",
    "print(f\"peptide precision:         {peptide_precision_refined:.5f}\")\n",
    "print(f\"peptide recall:            {peptide_recall_refined:.5f}\")\n",
    "print(f\"area under the ROC curve:  {auc_refined:.5f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Decrease in AA error rate:     {100 * (aa_error_rate - aa_error_rate_refined):.2f}%\")\n",
    "print(f\"Increase in AA precision:      {100 * (aa_precision_refined - aa_precision):.2f}%\")\n",
    "print(f\"Increase in AA recall:         {100 * (aa_recall_refined - aa_recall):.2f}%\")\n",
    "print(f\"Increase in peptide precision: {100 * (peptide_precision_refined - peptide_precision):.2f}%\")\n",
    "print(f\"Increase in peptide recall:    {100 * (peptide_recall_refined - peptide_recall):.2f}%\")\n",
    "print(f\"Increase in AUC:               {100 * (auc_refined - auc):.2f}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_predictions = pd.DataFrame(\n",
    "    {\n",
    "        \"targets\": targs,\n",
    "        \"tokenized_predictions\": predictions,\n",
    "        \"predictions\": [\"\".join(x) for x in predictions],\n",
    "        \"log_probabilities\": log_probs,\n",
    "    }\n",
    ")\n",
    "diffusion_predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_predictions.to_csv(\"diffusion_predictions.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
